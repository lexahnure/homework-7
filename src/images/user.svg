<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M6 8C6.5 8.96552 7 9.68967 8 9.68967V9.72104C7.22316 9.74106 6.5 9.5 6.5 9.5V10.5C4 11 2 12.2386 2 13.5517V15H8H14V13.5517C14 12.2386 12 11 9.5 10.5V8.84192C9.67781 8.59711 9.83891 8.31107 10 8C10.3966 7.23422 10.4786 7.03241 10.4956 6.5H10.5C10.679 6.5 10.7298 6.5 10.7442 6.47708C10.75 6.46799 10.75 6.45526 10.75 6.4375C10.75 6.375 10.75 6.25 11 6L11 5.5C11 5 10.5 5 10.5 5C10.5 3 10 2 8 2C6 2 5.5 3 5.5 5C5.5 5 5 5 5 5.5L5.00002 6C5.25003 6.25 5.25003 6.375 5.25002 6.4375C5.25002 6.45526 5.25002 6.46799 5.25575 6.47708C5.2702 6.5 5.32104 6.5 5.50002 6.5H5.50443C5.52139 7.03241 5.60344 7.23422 6 8Z" fill="#292E33"/>
</svg>
