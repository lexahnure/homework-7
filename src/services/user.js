import { rest } from './rest';

const login = data => rest.post(`public/login`, data);
const register = data => rest.post(`public/user`, data);

const checkUser = () => rest.get(`public/checkUser`);
const logout = () => rest.get('logout');
const updateUser = data => rest.put('user', data);

export { login, checkUser, register, logout, updateUser };
