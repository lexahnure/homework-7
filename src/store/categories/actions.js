export const GET_INFO_ASYNC = 'Get info';
export const getInfoAsync = () => ({ type: GET_INFO_ASYNC });

export const SET_INFO = 'Set info';
export const setInfo = data => ({ type: SET_INFO, data });

export const CLEAN_INFO = 'Clean info';
export const cleanInfo = () => ({ type: CLEAN_INFO });

export const GET_CATEGORIES_ASYNC = 'Get categories';
export const getCategoriesAsync = () => ({ type: GET_CATEGORIES_ASYNC });

export const SET_CATEGORIES = 'Set categories';
export const setCategories = data => ({ type: SET_CATEGORIES, data });

export const CLEAN_CATEGORIES = 'Clean categories';
export const cleanCategories = () => ({ type: CLEAN_CATEGORIES });

export const UPDATE_CATEGORIES_ASYNC = 'Update categories';
export const updateCategoriesAsync = data => ({ type: UPDATE_CATEGORIES_ASYNC, data });

export const GET_ONE_CATEGORY_ASYNC = 'Get one category async';
export const getOneCategoryAsync = data => ({ type: GET_ONE_CATEGORY_ASYNC, data });

export const SET_ONE_CATEGORY = 'Set one category to store';
export const setOneCategory = data => ({ type: SET_ONE_CATEGORY, data });

export const CLEAN_ONE_CATEGORY = 'Clean category in store';
export const cleanOneCategory = () => ({ type: CLEAN_ONE_CATEGORY });

export const UPDATE_ONE_CATEGORY_ASYNC = 'Update one category';
export const updateOneCategoryAsync = data => ({ type: UPDATE_ONE_CATEGORY_ASYNC, data });
