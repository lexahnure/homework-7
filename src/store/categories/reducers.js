import { SET_INFO, CLEAN_INFO, CLEAN_CATEGORIES, SET_CATEGORIES, SET_ONE_CATEGORY, CLEAN_ONE_CATEGORY } from './actions';
import { info as initial, categories as initialCategories, category as initialCategory } from '../initialState';

export const info = (state = initial, action) => {
  switch (action.type) {
    case SET_INFO: {
      return action.data;
    }
    case CLEAN_INFO: {
      return initial;
    }
  }
  return state;
};

export const categories = (state = initial, action) => {
  switch (action.type) {
    case SET_CATEGORIES: {
      return action.data;
    }
    case CLEAN_CATEGORIES: {
      return initialCategories;
    }
  }
  return state;
};

export const category = (state = initialCategory, action) => {
  switch (action.type) {
    case SET_ONE_CATEGORY: {
      return action.data;
    }
    case CLEAN_ONE_CATEGORY: {
      return initialCategory;
    }
  }
  return state;
};
