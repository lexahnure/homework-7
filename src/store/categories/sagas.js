import { takeEvery, put, all } from 'redux-saga/effects';
import { setInfo, GET_INFO_ASYNC, setCategories, GET_CATEGORIES_ASYNC, UPDATE_CATEGORIES_ASYNC, getCategoriesAsync, setOneCategory, GET_ONE_CATEGORY_ASYNC, UPDATE_ONE_CATEGORY_ASYNC } from './actions';
import { getInfo, getCategories, updateCategory, getCategory } from '../../services';
import { info as initial } from '../initialState';

function* getInfos() {
  const info = yield getInfo();
  yield put(setInfo(info));
}

function* getCategoriesSaga() {
  const categories = yield getCategories();
  yield put(setCategories(categories));
}

function* getOneCategorySaga(action) {
  const category = yield getCategory(action.data);
  yield put(setOneCategory(category));
}

function* updCategoriesSaga(action) {
  yield updateCategory(action.data);
  yield put(getCategoriesAsync());
}

function* updOneCategorySaga(action) {
  const { data } = action;
  const category = yield updateCategory(data);
  // const category = yield getCategory(data.id);
  yield put(setOneCategory(category));
}

export function* watchCategories() {
  yield all([
    takeEvery(GET_INFO_ASYNC, getInfos),
    takeEvery(GET_CATEGORIES_ASYNC, getCategoriesSaga),
    takeEvery(UPDATE_CATEGORIES_ASYNC, updCategoriesSaga),
    takeEvery(GET_ONE_CATEGORY_ASYNC, getOneCategorySaga),
    takeEvery(UPDATE_ONE_CATEGORY_ASYNC, updOneCategorySaga),
  ]);
}
