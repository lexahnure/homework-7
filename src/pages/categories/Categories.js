import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getCategoriesAsync, updateCategoriesAsync, cleanCategories } from 'store/categories';
import ControlItems from 'components/controlItems';

const isPublished = category => category.published;
const notPublished = category => !category.published;

class Categories extends Component {
  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(getCategoriesAsync());
  }

  componentWillUnmount() {
    const { dispatch } = this.props;
    dispatch(cleanCategories());
  }

  unpublishCategory = (id) => {
    const { dispatch, categories } = this.props;
    const category = categories.find(category => category.id === id);
    category.published = false;
    dispatch(updateCategoriesAsync(category));
  }

  publishCategory = (id) => {
    const { dispatch, categories } = this.props;
    const category = categories.find(category => category.id === id);
    category.published = true;
    dispatch(updateCategoriesAsync(category));
  }

  updateCategories = (title, id) => {
    const { dispatch, categories } = this.props;
    const category = categories.find(category => category.id === id);
    category.title = title;
    dispatch(updateCategoriesAsync(category));
  }

  render() {
    const { categories, history } = this.props;

    return (
      <>
        <section>
          <h2>Categories</h2>
          {
            categories && (
              <ControlItems
                leftItems={categories.filter(isPublished)}
                rightItems={categories.filter(notPublished)}
                onChangeLeftItem={this.updateCategories}
                removeItem={this.unpublishCategory}
                addItem={this.publishCategory}
                history={history}
                rightColumnName="Unpublished"
                leftColumnName="Publsihed"
              />
            )
          }
        </section>
      </>
    );
  }
}

const mapState = state => ({
  categories: state.categories
});

export default withRouter(connect(mapState)(Categories));
