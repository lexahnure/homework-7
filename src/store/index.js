import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { user } from './user';
import { info, categories, category } from './categories';
import { product, products } from './products';
import { rootSaga } from './rootSaga';
import { error } from './status';
// import * as reducers from './reducers';

const rootReducers = combineReducers({
  user,
  info,
  products,
  product,
  error,
  categories,
  category,
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const sagaMiddleware = createSagaMiddleware();


const store = createStore(
  rootReducers,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);

export const runSaga = () => sagaMiddleware.run(rootSaga);

export default store;
