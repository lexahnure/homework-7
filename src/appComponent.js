import { Redirect, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { ToastContainer } from 'react-toastr';
import Header from './components/header/index';
import Main from './components/main/index';
import { getProducts, requestUpdateProduct, requestDeleteProduct } from './services';
import { Pages } from './pages/Pages';
import { check, logout } from './store/user';
import { getInfoAsync, cleanInfo } from './store/categories';
import { cleanError } from './store/status';


class AppComp extends Component {
  state = {
    loading: true,
    items: [],
  }

  componentDidMount() {
    const { dispatch } = this.props;
  /*checkUser()
    .then(user => this.setState({ loading: false, user }))
    .catch(() => this.setState({ loading: false }));*/
    dispatch(check());
    this.setState({ loading: false });
  }

  componentDidUpdate(prevProps) {
    const { dispatch, history, error, user } = this.props;

    if (!prevProps.user && user) {
      dispatch(getInfoAsync());
    }
    if (prevProps.user && !user) {
      history.push('/');
    }
    if (!prevProps.error && error) {
      this.container.error(
        error,
        'Error!'
      );
      dispatch(cleanError());
    }
  }

  onLogout = () => {
    const { dispatch } = this.props;
    dispatch(logout());
    dispatch(cleanInfo());
  }

  updateProduct = (data) => {
    requestUpdateProduct(data)
      .then(() => getProducts()
        .then(items => this.setState({ items })));
  }

  deleteProduct = (id) => {
    requestDeleteProduct(id)
      .then(() => getProducts()
        .then(items => this.setState({ items })));
  }

  render() {
    const {
      loading,
      items,
    } = this.state;

    const { user, info } = this.props;

    return (
      <>
        <Header
          user={user}
          info={info}
          onLogout={this.onLogout}
        />
        <Main
          loading={loading}
        >
          <Pages
            user={user}
            items={items}
            updateProduct={this.updateProduct}
            deleteProduct={this.deleteProduct}
          />
        </Main>
        <ToastContainer
          ref={ref => this.container = ref}
          className="toast-top-right"
        />
      </>
    );
  }
}

const mapStateToProps = ({ user, error, info }) => ({ user, error, info });

export default withRouter(connect(mapStateToProps)(AppComp));

export { AppComp };
