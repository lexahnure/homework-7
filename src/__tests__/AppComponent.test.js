import { shallow } from 'enzyme';
import { getInfoAsync, cleanInfo } from 'store/categories';
import { logout } from 'store/user';
import { AppComp } from '../appComponent';

describe('<AppComponent>', () => {
  it('should match to snapshot', () => {
    const wrapper = shallow(<AppComp dispatch={_ => _} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Header and Main', () => {
    const wrapper = shallow(<AppComp dispatch={_ => _} />);
    expect(wrapper.find('Header').length).toBe(1);
    expect(wrapper.find('Main').length).toBe(1);
  });

  it('should send action getInfoAsync() if user changed', () => {
    const fakeDispatch = jest.fn();
    const wrapper = shallow(<AppComp dispatch={fakeDispatch} />);
    const user = {};

    wrapper.setProps({ user });
    expect(fakeDispatch).toHaveBeenCalledWith(getInfoAsync());
  });

  it('should send action history() if user changed', () => {
    const fakeDispatch = jest.fn();
    const history = { push: jest.fn() };
    const user = {};
    const wrapper = shallow(<AppComp history={history} user={user} dispatch={fakeDispatch} />);

    wrapper.setProps({ user: null });
    expect(history.push).toHaveBeenCalledWith('/');
  });

  it('should set error in error.container if get props error', () => {
    const fakeDispatch = jest.fn();
    const err = 'error';
    const Container = () => ({ error: jest.fn() });
    const wrapper = shallow(<AppComp error={null} dispatch={fakeDispatch} />);
    let toastCont = wrapper.find('ToastContainer')[0];
    toastCont = Container;

    wrapper.setProps({ err });
    expect(Container.error).toHaveBeenCalledWith('error');
  });

  it('should call dispatch two times with functions as arguments', () => {
    const fakeDispatch = jest.fn();
    const user = {};
    const wrapper = shallow(<AppComp dispatch={fakeDispatch} user={user} />);

    wrapper.simulate('onLogout');
    expect(fakeDispatch).toHaveBeenCalledWith(logout());
    expect(fakeDispatch).toHaveBeenCalledWith(cleanInfo());
  });
});
