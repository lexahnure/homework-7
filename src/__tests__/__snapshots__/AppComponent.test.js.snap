// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppComponent> should match to snapshot 1`] = `
<Fragment>
  <Header
    onLogout={[Function]}
  />
  <Main
    loading={false}
  >
    <Pages
      deleteProduct={[Function]}
      items={Array []}
      updateProduct={[Function]}
    />
  </Main>
  <ToastContainer
    className="toast-top-right"
    id="toast-container"
    newestOnTop={true}
    onClick={[Function]}
    preventDuplicates={true}
    toastMessageFactory={[Function]}
    toastType={
      Object {
        "error": "error",
        "info": "info",
        "success": "success",
        "warning": "warning",
      }
    }
  />
</Fragment>
`;
